### PR TITLE
Enable update tests for federated secret controller

### PR DESCRIPTION
**Pull request by mwielgus**
Enable update tests in secret controller + add option to execute multiple iterations in the same run.

cc: @nikhiljindal @madhusudancs

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/37307)

### CODE REVIEW

File: federation/pkg/federation-controller/secret/secret_controller_test.go [hunk 1]
```diff
@@ -116,38 +133,46 @@ func TestSecretController(t *testing.T) {
...
-				assert.Equal(t, secret1.Name, updatedSecret2.Name)
-				assert.Equal(t, secret1.Namespace, updatedSecret.Namespace)
-				assert.True(t, secretsEqual(secret1, *updatedSecret2),
-					fmt.Sprintf("expected: %v, actual: %v", secret1, *updatedSecret2))
-	*/
+			return nil
+		}
+	}
+
+	// TODO: Uncomment this once we have figured out why this is flaky.
```
Comment (nikhiljindal) [inline federation/pkg/federation-controller/secret/secret_controller_test.go]: remove?

File: federation/pkg/federation-controller/secret/secret_controller_test.go [hunk 2]
```diff
@@ -183,13 +208,17 @@ func GetSecretFromChan(c chan runtime.Object) *api_v1.Secret {
...
+	retryInterval := 200 * time.Millisecond
 	err := wait.PollImmediate(retryInterval, timeout, func() (bool, error) {
 		obj, found, err := store.GetByKey(clusterName, key)
 		if !found || err != nil {
+			glog.Errorf("%s is not in the store", key)
 			return false, err
 		}
 		equal := secretsEqual(*obj.(*api_v1.Secret), *desiredSecret)
+		if !equal {
+			glog.Errorf("wrong content in the store expected:\n%v\nactual:\n%v\n", *desiredSecret, *obj.(*api_v1.Secret))
```
Comment (nikhiljindal) [inline federation/pkg/federation-controller/secret/secret_controller_test.go]: Well this is not an error. This is expected to fail until eventually it becomes equal.
Change this to Infof?

File: federation/pkg/federation-controller/secret/secret_controller_test.go [hunk 3]
```diff
@@ -183,13 +208,17 @@ func GetSecretFromChan(c chan runtime.Object) *api_v1.Secret {
 
 // Wait till the store is updated with latest secret.
 func WaitForSecretStoreUpdate(store util.FederatedReadOnlyStore, clusterName, key string, desiredSecret *api_v1.Secret, timeout time.Duration) error {
-	retryInterval := 100 * time.Millisecond
+	retryInterval := 200 * time.Millisecond
 	err := wait.PollImmediate(retryInterval, timeout, func() (bool, error) {
 		obj, found, err := store.GetByKey(clusterName, key)
 		if !found || err != nil {
+			glog.Errorf("%s is not in the store", key)
```
Comment (nikhiljindal) [inline federation/pkg/federation-controller/secret/secret_controller_test.go]: Same. Change this to Infof?

File: federation/pkg/federation-controller/util/test/test_helper.go [hunk 4]
```diff
@@ -236,7 +236,7 @@ func CheckObjectFromChan(c chan runtime.Object, checkFunction CheckingFunction)
 			if lastError = checkFunction(obj); lastError == nil {
 				return nil
 			}
-			glog.Infof("Check function failed with %v", lastError)
+			glog.Errorf("Check function failed with %v", lastError)
```
Comment (nikhiljindal) [inline federation/pkg/federation-controller/util/test/test_helper.go]: Same. this should remain Infof

File: federation/pkg/federation-controller/secret/secret_controller_test.go [hunk 5]
```diff
@@ -34,10 +34,27 @@ import (
...
+	"github.com/golang/glog"
 	"github.com/stretchr/testify/assert"
 )
 
+const (
+	testIterations = 1
+)
+
 func TestSecretController(t *testing.T) {
+	for i := 0; i < testIterations; i++ {
```
Comment (nikhiljindal) [inline federation/pkg/federation-controller/secret/secret_controller_test.go]: Do you really want to check this in?

**Comment by nikhiljindal** [comment]
Can you confirm that this test is not flaky anymore?

File: federation/pkg/federation-controller/secret/secret_controller_test.go (hunk 5 above)
Comment (madhusudancs) [inline federation/pkg/federation-controller/secret/secret_controller_test.go]: Why do this at all? You could instead use [`cmd/stress`](https://godoc.org/golang.org/x/tools/cmd/stress) to accomplish this

```
$ go get -u  golang.org/x/tools/cmd/stress
$ go test -c pkg.test
$ stress ./pkg.test -test.run=TestMyFunc
```

**Comment by nikhiljindal** [comment]
Reassigning to @madhusudancs since I am out next week

File: federation/pkg/federation-controller/secret/secret_controller_test.go (hunk 5 above)
Comment (mwielgus) [inline federation/pkg/federation-controller/secret/secret_controller_test.go]: Hmm, interesting, didn't knew about this tool. Thanks.

Comment (mwielgus) [inline federation/pkg/federation-controller/secret/secret_controller_test.go]: Removed.

File: federation/pkg/federation-controller/secret/secret_controller_test.go (hunk 1 above)
Comment (mwielgus) [inline federation/pkg/federation-controller/secret/secret_controller_test.go]: Done.

File: federation/pkg/federation-controller/secret/secret_controller_test.go (hunk 3 above)
Comment (mwielgus) [inline federation/pkg/federation-controller/secret/secret_controller_test.go]: Done.

File: federation/pkg/federation-controller/secret/secret_controller_test.go (hunk 2 above)
Comment (mwielgus) [inline federation/pkg/federation-controller/secret/secret_controller_test.go]: Done.

File: federation/pkg/federation-controller/util/test/test_helper.go (hunk 4 above)
Comment (mwielgus) [inline federation/pkg/federation-controller/util/test/test_helper.go]: Done.

**Comment by mwielgus** [comment]
Yes, no more flakes spotted.

PTAL.

**Comment by dims** [comment]
@madhusudancs Do we need this in 1.5? please rebase! ( cc @matchstick )

**Comment by madhusudancs** [comment]
@dims not a release blocker. Can be cherry-picked after the release as well. So just leaving the milestone, but adding the `non-release-blocker` label.

**Review by madhusudancs — COMMENTED** [review]
@mwielgus this looks reasonable to me.

**Comment by madhusudancs** [comment]
@saad-ali for cherrypick approval. 

@saad-ali this is a test bug fix and contained within federation.

**Comment by saad-ali** [comment]
> madhusudancs commented an hour ago

Rebase needed!

**Pull request merged**